pkg/reconciler/tenancy/initialization/apibinder_initializer_controller.go

```diff
@@ -0,0 +1,204 @@
+/*
+Copyright 2022 The KCP Authors.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package initialization
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"time"
+
+	jsonpatch "github.com/evanphx/json-patch"
+	"github.com/kcp-dev/logicalcluster/v2"
+
+	"k8s.io/apimachinery/pkg/api/equality"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/clusters"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog/v2"
+
+	tenancyv1alpha1 "github.com/kcp-dev/kcp/pkg/apis/tenancy/v1alpha1"
+	kcpclient "github.com/kcp-dev/kcp/pkg/client/clientset/versioned"
+	tenancyinformer "github.com/kcp-dev/kcp/pkg/client/informers/externalversions/tenancy/v1alpha1"
+	tenancylister "github.com/kcp-dev/kcp/pkg/client/listers/tenancy/v1alpha1"
+	"github.com/kcp-dev/kcp/pkg/logging"
+)
+
+const (
+	apiBinderControllerName = "kcp-apibinder-initializer-controller"
+)
+
+// NewAPIBinder returns a new controller which instantiates APIBindings and waits for them to be fully bound
+// in new ClusterWorkspaces of any derivative type from APIBinder.
+func NewAPIBinder(
+	kcpClusterClient kcpclient.ClusterInterface,
+	clusterWorkspaceInformer tenancyinformer.ClusterWorkspaceInformer,
+) (*APIBinder, error) {
+	queue := workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), apiBinderControllerName)
+
+	c := &APIBinder{
+		initializer:            tenancyv1alpha1.ClusterWorkspaceAPIBindingsInitializer,
+		kcpClusterClient:       kcpClusterClient,
+		clusterWorkspaceLister: clusterWorkspaceInformer.Lister(),
+		queue:                  queue,
+	}
+
+	clusterWorkspaceInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			c.enqueueClusterWorkspace(obj)
+		},
+		DeleteFunc: func(obj interface{}) {
+			c.enqueueClusterWorkspace(obj)
+		},
+	})
+
+	return c, nil
+}
+
+// APIBinder is a controller which instantiates APIBindings and waits for them to be fully bound
+// in new ClusterWorkspaces.
+type APIBinder struct {
+	initializer tenancyv1alpha1.ClusterWorkspaceInitializer
```

Comment:
This is a hard-coded constant and can only ever be one thing, right? Seems like we'd be ok not having to hold a reference to it.